### ISSUE TITLE
Schema versioning and validation

### ISSUE THREAD

**Issue by yusefnapora** [mediachain/aleph#44]
Opening this to discuss and plan the schema semantics, especially versioning and translation.
Below are some notes that describe my current thinking:
## Versioning
- Schemas are stored as mediachain records in jsonschema format in a well-known namespace (e.g `mediachain.schemas` or something). 
- Schemas have human-readble names + semantic versions, which combined form the WKI used to look them up.  e.g. `SELECT * FROM mediachain.schemas WHERE wki = io.mediachain.indexer/image/jsonschema/1-0-0`
- A schema is "self-describing", in that it contains a json object that states its name, version, etc.
- Examples here use the snowplow [self-describing schema format](https://github.com/snowplow/iglu/wiki/Self-describing-JSON-Schemas) of `vendor/schema-name/format/version`, where version is a [SchemaVer](https://github.com/snowplow/iglu/wiki/SchemaVer) version string
- Mediachain records should contain a link to the schema that they conform to.
- Link format is up for discussion; actual IPLD-compatible multiaddrs don't support custom protocol codes, so linking directly to a mediachain object blob by multihash would probably need something else.
- I think we should probably have a "link object" that has, at minimum, the WKI of the schema + a multihash of the schema blob.  It could also have e.g. http links as a fallback.

Possible example of record with schema link:

```
{
    "schema": 
    {
        "wki": "io.mediachain.indexer/image/jsonschema/1-0-0," 
        "object": "QmF00...", 
        "http": "http://indexer.mediachain.io/schemas/image/jsonschema/1-0-0", 
        "ipfs": {"/": "QmF00"}
    },
    "data":
    {
        "id": "foo",
        "description": "actual record that conforms to schema above"
    }
}
```

The example above uses the [snowplow convention](https://github.com/snowplow/iglu/wiki/Self-describing-JSONs) of wrapping the actual payload in a `data` object, which lets us avoid having to designate a special "schema key" that you're not allowed to use in your actual data payload.

When a record is published, the schema object will be listed in the `deps` field of the statement, so that when you fetch the record you'll also get its schema.  This should ensure that we don't end up with "orphan" records whose schema can't be retrieved.  It might also be useful to put the WKI of the schema into the statement's `tags` field, so that users can filter by objects they know how to work with.
## Translation

A "translator" is a [jq filter](https://stedolan.github.io/jq/manual/#Basicfilters) that will massage data from your schema into another.
- A translator is published as a mediachain record; it can be a json object with links to the source and destination schemas, plus the jq filter string (and version of jq used!).
- translators are "one way", and may be lossy; there's no guarantee that the translated output can be re-translated back into the original schema
- However, a translated record should contain a link to the original record and the translator used.  This could be another top-level field like `schema` and `data` above
- In theory, translation can happen "on the fly" at read time, instead of a node republishing translated records.  Whether that's a good idea will depend on read volumes, etc.

**Comment by vyzo** [comment]
Let's simplify the schema pointer to the most primitive representation; I think we should aim to keep our objects as lean as possible.

We don't need to include anything ipfs related, neither do we need a url if it is stored in the datastore and fetched along with the metadata (by virtue of being listed in deps, already implemented in my wip branch).

So let's simplify the schema pointer to:
```
"schema": { "wki" : "...", "/" : "Qm..."}
```

Perhaps we don't even need the wki, it can be embedded in the schema object:
```
"schema": {"/": "Qm..."}
```

**Comment by vyzo** [comment]
Also, let's not abuse tags for schema purposes! 
Besides, the schema will already be listed in the object dependencies (albeight by hash)

**Comment by yusefnapora** [comment]
That sounds good to me; for something that's going to be attached to every record, these should be as light as we can manage.

Maybe instead of spamming tags, aleph nodes will maintain a nice index of schema hashes -> WKIs, so that you can easily search and filter by WKI, but we only need to publish the hash in the deps.

**Comment by vyzo** [comment]
Yeah, aleph should keep some local state to simplify the llife of the users.

So let's go for plain object pointer as schema, it's the leanest repr possible for our object representation:
```
{"schema": {"/": "Qm..."}, "data": {...}}
```

**Comment by vyzo** [comment]
We can even get creative and use single character keys for our two fields :)

**Comment by yusefnapora** [comment]
heh, I'm inclined to spend the eight bytes to keep the human readable names 😄 .  But I don't feel super strongly about it.  Definitely agree about the minimal representation.  If we do the IPLD transform, we can have nice compact multihash links with the link tag, so that will save us a few bytes as well.

**Comment by autoencoder** [comment]
One alternative to JSON-defined schema to consider is a flat schema definition, e.g:

- [Flattening (shredding) and reassembling nested JSON schemas -- Section 4 of the Dremel Paper](http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/36632.pdf)

- [SQL-like schema definition via REPEATED -- which internally uses the Dremel method](https://eventql.io/documentation/tables/datatypes/) and [see also](https://github.com/eventql/eventql/blob/master/doc/getting-started/concepts.md).

Benefits:

- Can be directly used to easily allow efficient storage of records in more traditional flat schema DBs, e.g. traditional columnar data warehouse DBs.

- Makes it easy to do all kinds of reasoning about the schema with fewer LOC.

**Comment by parkan** [comment]
A few points:

* referring to schemas by hash only makes semantic versioning useless (or requires complex expansion on the client), I would advocate for keeping the string around in the statement somewhere. I think it's a _reasonably_ safe assumption that a multi-statement will have a consistent schema (we can require this without much loss of generality)
* need to think about our commitment to IPLD. I think the slow progress on IPLD rollout has eroded our attention a bit, but that doesn't mean we should dismiss it altogether, there's definitely strength in collaboration
* most other points sound good, though let's not overspecify jq as the only translator format -- we should be able to represent chains of provenance and translation that happen outside mediachain or on non-JSON data 
* @autoencoder are you advocating specifically for columnar storage?

**Comment by autoencoder** [comment]
No :) Thinking about what the translation & schema definitions aim to accomplish, for who, and then how to optimize those cases.

**Comment by vyzo** [comment]
@parkan a couple of points:
- The idea is that that aleph will already auto-expand the schema object, so given that the complexity is there, we might as well save space on boilerplate. In terms of statement-level representation, the schema object will be listed as dependency and automatically merged.
- We are not giving up our commitment to IPLD -- our objects are serialized in IPLD format (CBOR + tags) and use the same hash format for references. When IPLD eventually materializes in IPFS, we can figure out how to hook our datastore to an ipfs node for making the metadata transparently accessible through IPFS. We will have to somehow solve the provider record problem however.

**Comment by yusefnapora** [comment]
So, there's a few things about the [schema branch](https://github.com/mediachain/aleph/pull/46) that feel awkward to me, and make the UX worse... 

- You need to manually define a schema up front before you can ingest anything.  Would be great if we can autogenerate a schema that fits your data and let you tweak it.

- There's no documentation yet about how schemas are defined, especially the "self-describing" bits I lifted from Snowplow.  So even if you have a valid json-schema, it's not obvious how to add the `self` property we require.  This could be addressed by the (hypothetical) schema generation step, where we could ask you to fill in the fields we need to populate the self-description.

- There's not yet a repository of "common" schemas that we can offer for mediachain-first data.  These should ideally be included in every copy of aleph, or hosted somewhere stable where we can fetch them at runtime.

- Referring to schemas by hash is awkward and confusing.  
	- At minimum I need to change the `schemaReference` argument to be a "named" flag, so you'd pass in `--schemaReference <hash>` instead of just the raw hash.  
	- I think it would also be nice if you could pass in a path to the schema file to the `publish` command instead of a hash, and it will hash it locally, try to fetch from concat, and publish if its missing.  Would add a few milliseconds, but we can do it the fast & cryptic way for bulk ingestions.

- Because schemas are attached to individual records instead of collections/namespaces, you can't easily point your code at a big batch of records and be confident that you know how to work with all the data inside.  This could lead to over-fetching records you don't know what to do with.
	- In theory, it's possible but awkward to make a MCQL query that returns statments whose object uses a specific schema, by asking for statements whose `deps` field contains the schema hash.  In practice, the `deps` field isn't yet exposed to mediachain queries, and is only used in the merge logic to fetch dependencies.  This also ignores schemas that are not exact matches but semver-compatible (as we discussed above).


I think that the schema generator is probably a good place to start; it could be an interactive command similar to `npm init` or something that will ask you to fill in a name, vendor and version for your schema.  You could point it at either an existing jsonschema or a directory full of json.  

That said, the PR is getting pretty huge as it is, so I don't know if we want to be stuffing more into it now.  I can focus on writing up some documentation of things as they are (and make the `--schemaReference` flag more explicit), and we can make it friendlier in a second pass.

**Comment by parkan** [comment]
Re: default schemas, see https://github.com/mediachain/aleph/issues/54

I agree with the awkwardness of using a hash in `deps`, see above, yeah.

Let's try to wrap this PR -- if schema validation is default-off/optional we should be ok. No need to document current state of things too much, let's get it right first.

**Comment by yusefnapora** [comment]
cool, that sounds good.  Right now `schemaReference` is required, even if you're not validating against it, since we're always making a self-describing record with a link to the schema.  But we can relax that and make the schema ref optional, in which case we just publish a bare record and let God sort it out 😄

I think that's probably a decent idea while we're still hashing out the details.

**Comment by parkan** [comment]
Yeah, shouldn't keep it that way for long but let's draw the line here right now

**Comment by vyzo** [comment]
I think it is simple to add the schema wki in the Statement envelope, as long as we don't need to index it or query it.

Implementation wise, we can extend the Statement message protobuf to include the schema wki with a new field (with a tag after the signature unfortunately), without invalidating existing statements.
The downside is that all objects in the statement must have the same schema. This is perhaps a reasonable assumption for compound statements, but it severly restricts republished envelope statements.

If we also need to query on schema, we will have to add the schema wki in the envelope table (and associated MCQL accessors). This will invalidate existing database and require migration, so it's a breaking chage. However, since existing statements wont have a schema, this can be a simple `ALTER TABLE`, so it's relatively painless.

In short, having schema wkis in the statement envelope is reasonable, and the sooner we do it the less painless it will be. The complication with envelope statements should be noted however.

**Comment by parkan** [comment]
Hmm why do you expect it to be a big issue for republishing?

**Comment by vyzo** [comment]
Just for republishing multiple statements together (perhaps for compression), not a problem in single statement republishing.

**Comment by yusefnapora** [comment]
Closing as we've got the first pass merged in.  Will open a new issue as things progress
